### ISSUE TITLE
[NEURON] no boards detected after apt-get upgrade

### ISSUE THREAD

**Issue by tonytee** [UniPiTechnology/evok#51]
### Prerequisites

* [ v] Are you running the latest Master or the latest release version?
* [ v] Did you try to reinstall EVOK?
* [ v] Did you perform a cursory search on the [forum] and [google]?

### Log files needed

Include the output of the following commands on your Neuron/UniPi device:

```
sudo su
cat /var/log/evok.access.log
ps -fax
cat /var/log/evok.log
```
root@S103-sn488:~# cat /var/log/evok.access.log
172.16.129.31 - - [21/Mar/2018:08:49:50 +0000] "GET /rest/all HTTP/1.1" 200 2644 "http://172.16.129.32/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/65.0.3325.162 Safari/537.36"
172.16.129.31 - - [21/Mar/2018:08:50:14 +0000] "GET /ws HTTP/1.1" 101 6570 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/65.0.3325.162 Safari/537.36"
172.16.129.31 - - [21/Mar/2018:08:50:14 +0000] "GET / HTTP/1.1" 304 0 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/65.0.3325.162 Safari/537.36"
172.16.129.31 - - [21/Mar/2018:08:50:15 +0000] "GET /rest/all HTTP/1.1" 304 0 "http://172.16.129.32/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/65.0.3325.162 Safari/537.36"

root@S103-sn488:~# ps -fax
  PID TTY      STAT   TIME COMMAND
    2 ?        S      0:00 [kthreadd]
    3 ?        S      0:00  \_ [ksoftirqd/0]
    5 ?        S<     0:00  \_ [kworker/0:0H]
    6 ?        S      0:00  \_ [kworker/u8:0]
    7 ?        S      0:00  \_ [rcu_sched]
    8 ?        S      0:00  \_ [rcu_bh]
    9 ?        S      0:00  \_ [migration/0]
   10 ?        S<     0:00  \_ [lru-add-drain]
   11 ?        S      0:00  \_ [cpuhp/0]
   12 ?        S      0:00  \_ [cpuhp/1]
   13 ?        S      0:00  \_ [migration/1]
   14 ?        S      0:00  \_ [ksoftirqd/1]
   16 ?        S<     0:00  \_ [kworker/1:0H]
   17 ?        S      0:00  \_ [cpuhp/2]
   18 ?        S      0:00  \_ [migration/2]
   19 ?        S      0:00  \_ [ksoftirqd/2]
   21 ?        S<     0:00  \_ [kworker/2:0H]
   22 ?        S      0:00  \_ [cpuhp/3]
   23 ?        S      0:00  \_ [migration/3]
   24 ?        S      0:00  \_ [ksoftirqd/3]
   26 ?        S<     0:00  \_ [kworker/3:0H]
   27 ?        S      0:00  \_ [kdevtmpfs]
   28 ?        S<     0:00  \_ [netns]
   29 ?        S      0:00  \_ [khungtaskd]
   30 ?        S      0:00  \_ [oom_reaper]
   31 ?        S<     0:00  \_ [writeback]
   32 ?        S      0:00  \_ [kcompactd0]
   33 ?        S<     0:00  \_ [crypto]
   34 ?        S<     0:00  \_ [bioset]
   35 ?        S<     0:00  \_ [kblockd]
   36 ?        S<     0:00  \_ [watchdogd]
   37 ?        S      0:00  \_ [kworker/0:1]
   38 ?        S<     0:00  \_ [rpciod]
   39 ?        S<     0:00  \_ [xprtiod]
   40 ?        S      0:00  \_ [kswapd0]
   41 ?        S<     0:00  \_ [vmstat]
   42 ?        S<     0:00  \_ [nfsiod]
   52 ?        S<     0:00  \_ [kthrotld]
   53 ?        S<     0:00  \_ [bioset]
   
   76 ?        S<     0:00  \_ [bioset]
   77 ?        S<     0:00  \_ [iscsi_eh]
   78 ?        S<     0:00  \_ [dwc_otg]
   79 ?        S      0:00  \_ [kworker/2:1]
   80 ?        S<     0:00  \_ [DWC Notificatio]
   81 ?        S<     0:00  \_ [VCHIQ-0]
   82 ?        S<     0:00  \_ [VCHIQr-0]
   83 ?        S<     0:00  \_ [VCHIQs-0]
   84 ?        S      0:00  \_ [VCHIQka-0]
   85 ?        S<     0:00  \_ [SMIO]
   87 ?        S      0:00  \_ [kworker/3:2]
   88 ?        S      0:00  \_ [irq/92-mmc1]
   90 ?        S<     0:00  \_ [bioset]
   91 ?        S      0:00  \_ [mmcqd/0]
   92 ?        S      0:00  \_ [kworker/1:1]
   93 ?        S      0:00  \_ [jbd2/mmcblk0p2-]
   94 ?        S<     0:00  \_ [ext4-rsv-conver]
   96 ?        S<     0:00  \_ [ipv6_addrconf]
  119 ?        S<     0:00  \_ [kworker/3:1H]
  127 ?        S      0:00  \_ [kworker/1:2]
  216 ?        S      0:00  \_ [spi0]
  248 ?        S<     0:00  \_ [cfg80211]
  250 ?        S<     0:00  \_ [brcmf_wq/mmc1:0]
  251 ?        S      0:00  \_ [brcmf_wdog/mmc1]
  265 ?        S      0:00  \_ [kworker/u8:2]
  275 ?        S      0:00  \_ [kworker/0:2]
  485 ?        S<     0:00  \_ [kworker/0:1H]
  486 ?        S<     0:00  \_ [kworker/1:1H]
  749 ?        S      0:00  \_ [kworker/2:0]
  750 ?        S      0:00  \_ [kworker/3:1]
  752 ?        S      0:00  \_ [kworker/0:0]
  754 ?        S      0:00  \_ [kworker/u8:1]
  779 ?        S      0:00  \_ [kworker/3:0]
    1 ?        Ss     0:03 /sbin/init
  117 ?        Ss     0:00 /lib/systemd/systemd-journald
  155 ?        Ss     0:00 /lib/systemd/systemd-udevd
  310 ?        Ssl    0:00 /lib/systemd/systemd-timesyncd
  361 ?        Ss     0:00 avahi-daemon: running [unipi-stretch.local]
  392 ?        S      0:00  \_ avahi-daemon: chroot helper
  363 ?        Ss     0:00 /lib/systemd/systemd-logind
  365 ?        Ssl    0:00 /usr/sbin/rsyslogd -n
  375 ?        Ss     0:00 /usr/sbin/cron -f
  377 ?        Ss     0:00 /usr/sbin/thd --triggers /etc/triggerhappy/triggers.d/ --socket /run/thd.socket --user nobody --deviceglob /dev/input/event*
  384 ?        Ss     0:00 /usr/bin/dbus-daemon --system --address=systemd: --nofork --nopidfile --systemd-activation
  398 ?        Ss     0:00 /opt/neuron-bin/neuron_tcp_server -p 502
  450 tty1     Ss+    0:00 /sbin/agetty --noclear tty1 linux
  453 ?        Ss     0:00 /sbin/netplugd -p /var/run/netplugd.pid
  455 ?        Ss     0:00 /usr/sbin/sshd -D
  649 ?        Ss     0:00  \_ sshd: pi [priv]
  670 ?        S      0:00      \_ sshd: pi@pts/0
  673 pts/0    Ss     0:00          \_ -bash
  690 pts/0    S      0:00              \_ sudo su
  694 pts/0    S      0:00                  \_ su
  699 pts/0    S      0:00                      \_ bash
  789 pts/0    R+     0:00                          \_ ps -fax
  460 ?        Ss     0:00 nginx: master process /usr/sbin/nginx -g daemon on; master_process on;
  461 ?        S      0:00  \_ nginx: worker process
  462 ?        S      0:00  \_ nginx: worker process
  463 ?        S      0:00  \_ nginx: worker process
  464 ?        S      0:00  \_ nginx: worker process
  540 ?        Ss     0:00 /sbin/dhclient -4 -v -pf /run/dhclient.eth0.pid -lf /var/lib/dhcp/dhclient.eth0.leases -I -df /var/lib/dhcp/dhclient6.eth0.leases eth0
  660 ?        Ss     0:00 /lib/systemd/systemd --user
  663 ?        S      0:00  \_ (sd-pam)
  778 ?        Ssl    0:04 /usr/bin/python /opt/evok/evok.py
  784 ?        S      0:00  \_ /usr/bin/python /opt/evok/evok.py

root@S103-sn488:/home/pi# python /opt/evok/evok.py
[I 180321 08:24:14 config:63] eprom: UniPi Neuron S103 version: 1.0 serial: 0x1e8
[I 180321 08:24:14 evok:1628] Starting using config file /etc/evok.conf
[I 180321 08:24:14 config:75] YAML Definition loaded: L20x.yaml, type: 3, definition count 0
[I 180321 08:24:14 config:75] YAML Definition loaded: M30x.yaml, type: 3, definition count 1
[I 180321 08:24:15 config:75] YAML Definition loaded: L50x.yaml, type: 3, definition count 2
[I 180321 08:24:15 config:75] YAML Definition loaded: S10x.yaml, type: 3, definition count 3
[I 180321 08:24:15 config:75] YAML Definition loaded: xS40.yaml, type: 3, definition count 4
[I 180321 08:24:15 config:75] YAML Definition loaded: M10x.yaml, type: 3, definition count 5
[I 180321 08:24:15 config:75] YAML Definition loaded: L30x.yaml, type: 3, definition count 6
[I 180321 08:24:15 config:75] YAML Definition loaded: M50x.yaml, type: 3, definition count 7
[I 180321 08:24:16 config:75] YAML Definition loaded: L51x.yaml, type: 3, definition count 8
[I 180321 08:24:16 config:75] YAML Definition loaded: xS30.yaml, type: 3, definition count 9
[I 180321 08:24:16 config:75] YAML Definition loaded: M40x.yaml, type: 3, definition count 10
[I 180321 08:24:16 config:75] YAML Definition loaded: M20x.yaml, type: 3, definition count 11
[I 180321 08:24:16 config:75] YAML Definition loaded: CUSTOM MODBUS DEVICE.yaml, type: 3, definition count 12
[I 180321 08:24:16 config:75] YAML Definition loaded: L40x.yaml, type: 3, definition count 13
[I 180321 08:24:17 config:75] YAML Definition loaded: xS10.yaml, type: 3, definition count 14
[I 180321 08:24:17 config:75] YAML Definition loaded: xS50.yaml, type: 3, definition count 15
[I 180321 08:24:17 config:75] YAML Definition loaded: evok-alias.yaml, type: 2, definition count 0
[I 180321 08:24:17 evok:1692] HTTP server listening on port: 8080
[I 180321 08:24:17 modbusclient_tornado:122] SPI client started
[I 180321 08:24:17 neuron:157] Reading SPI boards
[I 180321 08:24:17 neuron:172] No board on SPI 1
[I 180321 08:24:17 neuron:172] No board on SPI 2
[D 180321 08:24:17 owclient:377] Entering 1wire loop
[I 180321 08:24:17 neuron:172] No board on SPI 3
[I 180321 08:24:17 web:2063] 101 GET /ws (127.0.0.1) 3.58ms
[D 180321 08:24:17 evok:158] New WebSocket client connected
[I 180321 08:24:17 web:2063] 304 GET /rest/all (127.0.0.1) 4.17ms
[E 180321 08:24:17 neuron:211] 'NoneType' object has no attribute 'do_scan'
Traceback (most recent call last):
File "/opt/evok/neuron.py", line 209, in scan_boards
yield self.modbus_cache_map.do_scan()
AttributeError: 'NoneType' object has no attribute 'do_scan'

### Description

evok "breaks" after apt-get upgrade

### Steps to Reproduce

1. start out with image UniPian-Neuron-OS-2018-03-09.img all works fine out of the box
Evok fine everything fine

=> Linux S103-sn488 4.9.41-v7+ #1023 SMP Tue Aug 8 16:00:15 BST 2017 armv7l GNU/Linux

2. apt-get update / apt-get upgrade

=> Linux S103-sn488 4.9.80-v7+ #1098 SMP Fri Mar 9 19:11:42 GMT 2018 armv7l GNU/Linux

result: No boards found on SPI 1, 2, 3    
(problem arises after reboot, not before, so probably because of new kernel)

3. tried manual install of git evok-2.0.3a didn't fix things

**Expected behavior:** expected to be able to run apt-get upgrade safely

**Actual behavior:** evok doesn't find board on SPI anymore and serves out empty dashboard

**long shot:** possibly connected to https://github.com/raspberrypi/linux/commit/3b1e49ba1a8e5853c75914eee12000aa18ada46d ??

[forum]:https://forum.unipi.technology/category/4/official-evok-api
-- posted to forum too, sorry if doubles are inconvenient

[google]:http://www.google.com/

cheers Tony

**Comment by tknot** [comment]
As with my reply on the forum, it seems likely that the kernel driver module (either neuronspi.ko or unipi.ko, depending on which distribution you have) is unable to load due to the RPi kernel update performed by apt-get upgrade. I'll test this myself further, but you _may_ be able to fix this by  moving the aforementioned file (either unipi.ko or neuronspi.ko, depending on the image version you have) from 
`
/lib/modules/[the oldest version in the folder]-v7+/kernel/drivers/spi/
`
to
`
/lib/modules/[the newest version in the folder]-v7+/kernel/drivers/spi/
`

This isn't guaranteed to work as I haven't tested it with the latest update to the kernel, but it is likely to.

More generally, this issue arises from the fact that linux kernel drivers are always compiled against a specific kernel version, unlike e.g. on windows. Most universal drivers come included with the kernel, while large corporations such as e.g. NVidia create a binary version for each kernel version which gets included in common distributions. This is unfortunately not feasible for us.

Within the Linux community the "_intended_" approach is to compile the kernel module yourself; this is however not entirely trivial. You can find the source code for our kernel modules at [http://git.unipi.technology:3000/UniPi/neuron-kernel](http://git.unipi.technology:3000/UniPi/neuron-kernel)

**Comment by tknot** [comment]
We should be able to fix this by creating a dummy .deb package dependent on the kernel, which will prevent it from being updated by apt-get upgrade.

**Comment by tonytee** [comment]
Thanks for your reply, 

I have tried your suggestion by copying
/lib/modules/4.9.41-v7+/kernel/drivers/spi/neuronspi.ko
to both
/lib/modules/4.9.80-v7+/kernel/drivers/spi/
and
/lib/modules/4.9.80+/kernel/drivers/spi/

but sofar no fix... it could be that i broke too much on the image by now, so i need to try this again on a clean one, with a clean apt-get update/upgrade.

also, maybe some other .ko's are not playing nicely
in the newer spi folders next to the "old"
-rw-r--r-- 1 root root 41226 Mar 22 17:58 neuronspi.ko
 i also have
-rw-r--r-- 1 root root 14416 Mar 13 20:53 spi-bcm2835.ko
-rw-r--r-- 1 root root 11152 Mar 13 20:53 spi-bcm2835aux.ko
-rw-r--r-- 1 root root 16072 Mar 13 20:53 spidev.ko

if this quick fix neither works on the clean image i'll give the module compile a go.

i'll get back to you with my findings on both routes tomorrow...

thanks, tony

**Comment by yaneti** [comment]
> Within the Linux community the 'intended' approach is to compile the kernel module yourself; 

Nope , the intention is for all hardware manufacturers to see the light and get their kernel code upstream. ;)

**Comment by tknot** [comment]
Nope , the intention is for all hardware manufacturers to see the light and get their kernel code upstream. ;)

Point taken :) Our drivers don't really fit within the established framework of kernel drivers unfortunately though, some of it is due to business decisions (the multiple levels of platform devices for one) and some of it is due to the nature of the device. One could note that no other PLC manufacturer has drivers within the kernel, despite us knowing of several which use mainstream Linux.

And, well, they would need to be cleaned up quite a bit before they could get upstream.

**Comment by tonytee** [comment]
Hi Tomas,

Sorry to say that I can now confirm that the "moving neuronspi.ko to new location fix" does not work on a clean install either.

that is, starting out with UniPian-Neuron-OS-2018-03-09.img, seeying taht it actually works, and a apt-get update/upgrade right after that.

I haven't put much time in the module recompile, couldn't get it to work... as you said, not so trivial. 
At least not to me :)

My conclusion is that there's not an easy fix for rescuing a broken system, so just sheer providence made that i luckily didn't need that fix anywhere (yet)

However I also looked into your package pinning suggestion.

apt-mark hold raspberrypi-bootloader
apt-mark hold python-tornado
apt-mark hold raspberrypi-kernel

before an upgrade at least prevents things getting broken

so at least there's a temporary prevention scheme

cheers, please let me know if i can help more

tony

**Comment by tknot** [comment]
It looks like we'll have to pin the packages on our images as well, at least for now. Our .deb package repository is taking longer to setup than we anticipated, as we have decided to include CI and all of our other software as well.

We will try to keep the packages pinned by default in the next image version, and fix this properly once we have our repository up and running properly.

**Comment by tknot** [comment]
The package system rollout should happen within the next week or two, barring any unforeseen circumstances. Internal testing is going well, the package repository will be on files.unipi.technology.

**Comment by tknot** [comment]
As the package system is now active it should be possible to perform apt-get upgrade freely, provided that EVOK is installed via a debian package.

**Comment by daankets** [comment]
@tknot Today, I updated my UniPI Neuron to the latest raspberry pi kernel using `sudo apt-get upgrade`... The result:
```
The following packages have unmet dependencies:
 neuron-kernel : Depends: raspberrypi-kernel (= 1.20181112-1) but 1.20190401-1 is to be installed
E: Unable to correct problems, you have held broken packages.```

And therefore, Evok lists 0 devices (as the neuron kernel is missing). So it seems this issue is not yet resolved?

**Comment by tomashora** [comment]
That is a general dependency issue. After RPi releases a new kernel version, we have to test everything and release a new version of our kernel drivers which usually takes some time. If users install/update the RPi kernel before we release the new compatible kernel driver, the install fails which causes malfunctional Evok.
@daankets That is fixed after the release of neuron-kernel: 1.22.1.20190401-1 and unipi-kernel-modules-dkms: 1.22 Please apt update & upgrade.
